spring-cloud-kubernetes-client-discovery/src/main/java/org/springframework/cloud/kubernetes/client/discovery/KubernetesInformerDiscoveryClient.java

```diff
@@ -145,39 +159,61 @@ private Stream<ServiceInstance> getServiceInstanceDetails(V1Service service, Str
-								p -> metadata.put(StringUtils.hasText(p.getName()) ? p.getName() : UNSET_PORT_NAME,
-										Integer.toString(p.getPort())));
-					}
-					List<V1EndpointAddress> addresses = subset.getAddresses();
-					if (addresses == null) {
-						addresses = new ArrayList<>();
-					}
-					if (this.properties.includeNotReadyAddresses()
-							&& !CollectionUtils.isEmpty(subset.getNotReadyAddresses())) {
-						addresses.addAll(subset.getNotReadyAddresses());
-					}
-
-					final int port = findEndpointPort(endpointPorts, primaryPortName, serviceId);
-					return addresses.stream()
-							.map(addr -> new DefaultKubernetesServiceInstance(
-									addr.getTargetRef() != null ? addr.getTargetRef().getUid() : "", serviceId,
-									addr.getIp(), port, metadata, secured, service.getMetadata().getNamespace(),
-									// TODO find out how to get cluster name possibly from
-									// KubeConfig
-									null));
-				});
+		Map<String, V1ObjectMeta> finalPodMap = podMap;
+		return ep.getSubsets().stream()
+			.filter(subset -> subset.getPorts() != null && subset.getPorts().size() > 0) // safeguard
+			.flatMap(subset -> {
+				Map<String, String> metadata = new HashMap<>(svcMetadata);
+				List<CoreV1EndpointPort> endpointPorts = subset.getPorts();
+				if (this.properties.metadata() != null && this.properties.metadata().addPorts()) {
+					endpointPorts.forEach(
+						p -> metadata.put(StringUtils.hasText(p.getName()) ? p.getName() : UNSET_PORT_NAME,
+							Integer.toString(p.getPort())));
+				}
+				List<V1EndpointAddress> addresses = subset.getAddresses();
+				if (addresses == null) {
+					addresses = new ArrayList<>();
+				}
+				if (this.properties.includeNotReadyAddresses()
+					&& !CollectionUtils.isEmpty(subset.getNotReadyAddresses())) {
+					addresses.addAll(subset.getNotReadyAddresses());
+				}
+
+				final int port = findEndpointPort(endpointPorts, primaryPortName, serviceId);
+				return addresses.stream()
+					.map(addr -> new DefaultKubernetesServiceInstance(
+						addr.getTargetRef() != null ? addr.getTargetRef().getUid() : "", serviceId,
+						addr.getIp(), port, properties.initPodMetaData() ? initPodMetaData(addr, metadata, finalPodMap) : metadata, secured, service.getMetadata()
+						.getNamespace(),
+						// TODO find out how to get cluster name possibly from
+						// KubeConfig
+						null));
+
+			});
+	}
+
+	private Map<String, String> initPodMetaData(V1EndpointAddress addr, Map<String, String> metadata, Map<String, V1ObjectMeta> podMap) {
+		Map<String, String> newMetaData = new HashMap<>(metadata);
+		if (!CollectionUtils.isEmpty(podMap)) {
+			V1ObjectReference targetRef = addr.getTargetRef();
+			if (!ObjectUtils.isEmpty(targetRef)) {
+				String name = targetRef.getName();
+				if (!ObjectUtils.isEmpty(name)) {
+					Optional.ofNullable(podMap.get(name)).ifPresent(podMetaData -> {
+						if (!CollectionUtils.isEmpty(podMetaData.getLabels())) {
+							newMetaData.putAll(podMetaData.getLabels());
```

Comment:
is this just adding labels, what about annotations?

Reply:
The metadata obtained here is the metadata of the pod, which should belong to metadata by definition

Reply:
done